cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,540 @@
+    know the usual maximum needed for my workload (beyond it normal scale up
+    would apply)
+
+## User Stories (Optional)
+
+### Story 1
+
+The user has a deployment of pods that are serving user traffic. In order to
+better contain the traffic spikes I want to keep a buffer of a size of 10% of
+the deployment size so that when the HPA scales my deployment the new pods are
+started faster. Additionally, this speeds up the feedback loop of the metrics
+allowing the HPA also to faster provide next scaling decisions.
+
+Example buffer:
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: CapacityBuffer
+metadata:
+  name: my-deployment-buffer
+  namespace: my-namespace
+spec:
+  scalableRef
+    kind: Deployment
+    name: my-deployment
+  percentage: 10
+  replicas: 1
+```
+
+### Story 2
+
+The user is a part of an admin team owning the CI/CD pipeline. The cluster is
+used for running one off test tasks that should be completed promptly. However,
+the cluster is scaling up and down making the tasks wait for adding a new node.
+The admin team wants to keep spare space in the cluster so that the tasks
+schedule quickly.
+
+Example buffer:
+```
+apiVersion: v1
+kind: PodTemplate
+metadata:
+  name: my-custom-template
+  namespace: my-namespace
+template:
+  spec:
+    containers:
+    - name: cpu-buffer-container
+      resources:
+        requests:
+          cpu: "8"
+        limits:
+          cpu: "8"
+        nodeSelector:
+          cloud.google.com/compute-class: my-custom-class
+
+
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: CapacityBuffer
+metadata:
+  name: testing-capacity
+  namespace: my-namespace
+spec:
+  limits:
+    mem: 5120Mi
+    cpi: 40
+  podTemplateRef:
+    name: my-custom-template
+```
+
+## Notes/Constraints/Caveats (Optional)
+
+*   The NodeAutoscalingBuffer implementation assumes that the pods run as a part 
+    of scale subresource are homogeneous.
+
+## Risks and Mitigations
+
+*   The implementation of buffers will introduce additional “virtual” pods to be
+    handled by the autoscalers. In case the buffer chunks were very small the
+    number of processed pods may grow - we do not expect that buffers would
+    introduce more chunks than the number of pods that user may want to run.
+    Since these are only virtual they will introduce less overhead in the
+    cluster than balloon pods/deployments (no need to preempt etc).
+
+# Design Details
+
+## Proposed API
+
+```go
+type CapacityBuffer struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec   CapacityBufferSpec
+  Status CapacityBufferStatus
+}
+
+type LocalObjectRef struct {
+  Name string
+}
+
+type ScalableRef struct {
+  // optional
+  ApiGroup *string
+  // The first version advertised in API discovery for the specified apiGroup
+  // serving the specified kind with a scale subresource is used.
+  Kind string
+  Name string
+}
+
+type CapacityBufferSpec struct {
+  // optional
+  // oneof=PodShapeSource
+  PodTemplateRef *LocalObjectRef
```

Comment:
for `ScalableRef ` we have the comment:

`// Reference to an object of a kind that has scale subresource and sets its label selector field.`

Let's add a similar comment for `PodTemplateRef`, maybe something like:

`// Reference to a pre-defined PodTemplate resource in the same namespace that declares a static buffer`